src/rgw/driver/rados/oidc.cc

```diff
@@ -0,0 +1,285 @@
+        << " with: " << cpp_strerror(ret) << dendl;
+    return ret;
+  }
+
+  if (mdlog) {
+    return mdlog->complete_entry(dpp, y, "oidc", info.id, &objv);
+  }
+
+  return 0;
+}
+
+int remove(const DoutPrefixProvider* dpp, optional_yield y,
+           RGWSI_SysObj& sysobj, RGWSI_MDLog* mdlog,
+           const RGWZoneParams& zone, const std::string_view tenant,
+           std::string_view url,
+           RGWObjVersionTracker& objv,
+           std::string& id)
+{
+  std::string oid = oidc_provider_oid(tenant, oidc_url_oid_prefix, url);
+
+  int ret = rgw_delete_system_obj(dpp, &sysobj, zone.oidc_pool, oid, &objv, y);
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << "ERROR: deleting oidc provider url from pool: " << zone.oidc_pool.name << ": "
+                  << url << ": " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  if (mdlog) {
+    return mdlog->complete_entry(dpp, y, "oidc", id, &objv);
+  }
+
+  return 0;
+}
+
+
+class MetadataObject : public RGWMetadataObject {
+  RGWOIDCProviderInfo info;
+public:
+  MetadataObject(const RGWOIDCProviderInfo& info, const obj_version& v, real_time m)
+    : RGWMetadataObject(v, m), info(info) {}
+
+  void dump(Formatter *f) const override {
+    info.dump(f);
+  }
+
+  RGWOIDCProviderInfo& get_oidc_info() {
+    return info;
+  }
+};
+
+class MetadataLister : public RGWMetadataLister {
+ public:
+  using RGWMetadataLister::RGWMetadataLister;
+
+  virtual void filter_transform(std::vector<std::string>& oids,
+                                std::list<std::string>& keys) {
+    // remove the oid prefix from keys
+    constexpr auto trim = [] (const std::string& oid) {
+      return oid.substr(oid_prefix.size());
+    };
+    std::transform(oids.begin(), oids.end(),
+                   std::back_inserter(keys),
+                   trim);
+  }
+};
+
+class MetadataHandler : public RGWMetadataHandler {
+  RGWSI_SysObj& sysobj;
+  RGWSI_MDLog& mdlog;
+  const RGWZoneParams& zone;
+ public:
+  MetadataHandler(RGWSI_SysObj& sysobj, RGWSI_MDLog& mdlog,
+                  const RGWZoneParams& zone)
+    : sysobj(sysobj), mdlog(mdlog), zone(zone) {}
+
+  std::string get_type() final { return "oidc";  }
+
+  RGWMetadataObject* get_meta_obj(JSONObj *jo,
+                                  const obj_version& objv,
+                                  const ceph::real_time& mtime) override
+  {
+    RGWOIDCProviderInfo info;
+
+    try {
+      info.decode_json(jo);
+    } catch (JSONDecoder:: err& e) {
+      return nullptr;
+    }
+
+    return new MetadataObject(info, objv, mtime);
+  }
+
+  int get(std::string& entry, RGWMetadataObject** obj,
+          optional_yield y, const DoutPrefixProvider* dpp) override
+  {
+    RGWOIDCProviderInfo info;
+    std::string_view account;
+    std::string_view url;
+    int ret = read(dpp, y, sysobj,
+                   zone, account, url,
+                   info.id, info,
+                   &info.mtime, &info.objv_tracker);
```

Comment:
this is a request to read the piece of metadata identified by the `std::string& entry` argument, but we're not using `entry` at all here. that argument corresponds to the metadata key, so we need to define some mapping between these metadata keys and their rados oids

the oid format specified by `oidc_provider_oid()` is `{account}oidc.{url}`. for the metadata keys, i'd suggest using a delimiter like `$`, ie `{account}${url}`

then `MetadataHandler` needs to parse the `account` and `url` variables out of `entry` in order for calls like `read()` to work correctly

we also need to teach `MetadataLister` how to convert the oid strings back to their metadata keys